### DISCUSSION TITLE
Based on lightweight threads or native (OS) threads?

### DISCUSSION THREAD

**Discussion by monkeyjunglejuice** [mdbergmann/cl-gserver#97] [Ideas]
I'm currently evaluating various languages and runtimes to build a simulation that requires massive concurrency and task parallelism – millions (and later billions) of fluid, interacting compound entities. For now, everything seems to point into the BEAM/Elixir direction (GC, shared-nothing memory per process) – even though I would love to work interactively with CL instead of Elixir. Do Sento's actors map to native threads, or is there a (preemptive?) scheduler behind with lightweight processes? (Another actor framework I'm considering is https://riot.ml)

**Comment by mdbergmann** [comment]
Hello.

Sento actors don't by default map to native threads. That wouldn't scale. Instead Sento uses message-dispatchers which are kind of thread-pools of native threads.
By default each actor creates a mailbox where each message is processed by one of the threads of the message-dispatcher. The number of threads in the message-dispatcher is configurable. It's also possible to setup multiple message-dispatchers for different functional or business domains.
This effectively allows to create millions of actors. The way messages are handled by the message-dispatchers is currently done by a random or round-robin strategy, so a very simple way. Possible (but not implemented) other strategies are a more 'balanced' way by checking idle or queue sizes of actor mailboxes.

I'm working with Akka, a Java/Scala based Actor system, in production (professionally) which also 'just' has to work with native threads. But optimizing the right balance of number and size of message dispatchers and other worker thread-pool in large systems allows a massive IO throughput of many millions of asynchronous requests (or items of work).

Eventually also Erlang has to map all its virtual threads to native threads. But Erlang (OTP) runtime is really very well optimized for massive concurrent work.

**Reply by monkeyjunglejuice** [reply]
Ok, thank you for giving a starting point. Guess I'm going to play around with it :) I wasn’t quite sure how Sento’s actors compare to Erlang’s process model - it looks quite different at a first glance.
